File: python/false_feet_algorithm/examples/llm_reasoning_robustness.py
#!/usr/bin/env python3
"""
LLM Reasoning Robustness: Multi-Path Convergence Analysis

Uses the False Feet Algorithm concept to verify LLM reasoning quality:
- Generate multiple independent reasoning paths for the same problem
- Evaluate each path's internal consistency and strength
- Use INVERSE resource allocation (give more scrutiny to weaker paths)
- Check for convergence across paths → high confidence in reasoning

This addresses a critical problem: "How do I know my LLM's answer is actually robust?"
Answer: When multiple independent reasoning approaches converge, that's a strong signal.
"""

import json
import re
from dataclasses import dataclass
from typing import Optional


@dataclass
class ReasoningPath:
    """Represents one independent reasoning approach"""
    name: str
    prompt_style: str  # e.g., "step-by-step", "devil's advocate", "constraint-based"
    reasoning: str
    conclusion: str
    confidence: float = 0.0
    strength_score: float = 0.0
    iterations: int = 0


class LLMReasoningRobustness:
    """
    Evaluates LLM reasoning robustness through multi-path convergence.
    
    Core principle: When weak reasoning paths are forced to strengthen or die,
    and multiple paths converge on the same conclusion, that's a robust signal.
    """
    
    def __init__(self, problem: str):
        self.problem = problem
        self.paths: list[ReasoningPath] = []
        self.iteration = 0
        self.max_iterations = 5
        
    def create_reasoning_prompts(self) -> dict[str, str]:
        """Create multiple independent reasoning frameworks for the same problem"""
        return {
            "step_by_step": f"""Solve this step-by-step, breaking down the problem into clear stages:
Problem: {self.problem}
Format your answer as: Step 1 → Step 2 → Step 3 → Conclusion""",
            
            "devil's_advocate": f"""Take the opposite position first, then refute it to arrive at the true answer:
Problem: {self.problem}
Format: What if the opposite were true? Why would that fail? Therefore...""",
            
            "constraint_based": f"""Identify all constraints, then work within them:
Problem: {self.problem}
Format: Constraints are... Given these limits, the conclusion is...""",
            
            "evidence_first": f"""What evidence or data points matter most?
Problem: {self.problem}
Format: Key evidence: ... This means... Therefore...""",
        }
    
    def simulate_llm_response(self, prompt_style: str, iteration: int) -> str:
        """
        Simulate LLM response (in practice, call Claude API here)
        Shows what multi-path reasoning might look like
        """
        # Example: Complex reasoning problem
        # (In production, this would call Claude API via anthropic.Anthropic().messages.create())
        
        responses = {
            "step_by_step": [
                "Step 1: The problem requires analyzing multiple factors. Step 2: Factor A suggests X, Factor B suggests Y. Step 3: When combined with constraint Z, both point toward conclusion C.",
                "Step 1: Re-examining more carefully. Step 2: Factor A is stronger than initially thought. Step 3: This reinforces conclusion C."
            ],
            "devil's_advocate": [
                "What if the opposite were true? That would require ignoring Factor A entirely, which doesn't hold up. Therefore, conclusion C is correct.",
                "Devil's advocate: Could we justify the opposite? No—evidence contradicts it. Conclusion C stands."
            ],
            "constraint_based": [
                "Constraints: Time limit, budget, regulatory. Within these: Options A and B fail. Only conclusion C satisfies all constraints.",
                "Constraints confirmed: Conclusion C is the only option meeting all requirements."
            ],
            "evidence_first": [
                "Key evidence: Data shows 70% support for C. This dominates other signals. Conclusion: C.",
                "Evidence review: 70% signal for C remains primary. Conclusion C confirmed."
            ]
        }
        
        return responses.get(prompt_style, ["Unable to generate response"])[
            min(iteration, len(responses.get(prompt_style, [])) - 1)
        ]
    
    def extract_conclusion(self, reasoning: str) -> str:
        """Extract the core conclusion from reasoning"""
        # Simple pattern matching (in production, use Claude to extract)
        if "conclusion C" in reasoning.lower():
            return "C"
        elif "C is correct" in reasoning:
            return "C"
        elif "Conclusion: C" in reasoning:
            return "C"
        return "Unknown"
    
    def evaluate_path_strength(self, path: ReasoningPath) -> float:
        """
        Score a reasoning path's internal consistency and clarity.
        Returns 0.0-1.0 where higher = stronger.
        """
        score = 0.0
        
        # Strength signals
        if len(path.reasoning) > 50:
            score += 0.2
        if "therefore" in path.reasoning.lower() or "conclusion" in path.reasoning.lower():
            score += 0.3
        if path.conclusion in ["A", "B", "C", "D"]:
            score += 0.2
        if "constraint" in path.reasoning.lower() or "evidence" in path.reasoning.lower():
            score += 0.2
        if path.iterations > 1:
            score += 0.1  # Bonus for refinement
            
        return min(score, 1.0)
    
    def check_convergence(self) -> tuple[bool, str, float]:
        """
        Check if paths converge on the same conclusion.
        Returns: (converged: bool, conclusion: str, confidence: float)
        """
        if not self.paths:
            return False, "No paths evaluated", 0.0
        
        conclusions = [p.conclusion for p in self.paths]
        unique_conclusions = set(conclusions)
        
        if len(unique_conclusions) == 1:
            # Perfect convergence
            avg_strength = sum(p.strength_score for p in self.paths) / len(self.paths)
            confidence = 0.7 + (avg_strength * 0.3)  # Strength boosts confidence
            return True, conclusions[0], min(confidence, 1.0)
        else:
            # Partial convergence
            most_common = max(set(conclusions), key=conclusions.count)
            convergence_rate = conclusions.count(most_common) / len(conclusions)
            confidence = convergence_rate * 0.5  # Lower confidence if not unanimous
            return convergence_rate > 0.66, most_common, confidence
    
    def allocate_resources_inverse(self) -> dict[int, float]:
        """
        INVERSE resource allocation: Give MORE scrutiny to weaker paths.
        
        In False Feet, weak branches get more investigation slots.
        Here: weaker reasoning paths get flagged for more scrutiny.
        """
        if not self.paths:
            return {}
        
        strengths = [p.strength_score for p in self.paths]
        max_strength = max(strengths) if strengths else 1.0
        
        # Inverse: weaker = higher resource allocation
        allocations = {}
        for i, path in enumerate(self.paths):
            weakness = 1.0 - (path.strength_score / max(max_strength, 0.1))
            allocations[i] = 0.5 + (weakness * 0.5)  # Range 0.5-1.0
        
        return allocations
    
    def run_investigation(self) -> dict:
        """Execute multi-path reasoning investigation"""
        print(f"\n{'='*70}")
        print(f"LLM REASONING ROBUSTNESS INVESTIGATION")
        print(f"Problem: {self.problem}")
        print(f"{'='*70}\n")
        
        prompts = self.create_reasoning_prompts()
        
        # Initialize paths
        for name, prompt_style in prompts.items():
            self.paths.append(ReasoningPath(
                name=name,
                prompt_style=name,
                reasoning="",
                conclusion=""
            ))
        
        # Iterative investigation with inverse resource allocation
        for iteration in range(self.max_iterations):
            self.iteration = iteration
            print(f"\n--- ITERATION {iteration + 1} ---\n")
            
            # Evaluate current paths
            for path in self.paths:
                response = self.simulate_llm_response(path.prompt_style, iteration)
                path.reasoning = response
                path.conclusion = self.extract_conclusion(response)
                path.strength_score = self.evaluate_path_strength(path)
                path.iterations = iteration + 1
            
            # Check convergence
            converged, conclusion, confidence = self.check_convergence()
            
            print("Path Evaluations:")
            for i, path in enumerate(self.paths):
                print(f"  {i}. {path.name:20} | Conclusion: {path.conclusion} | Strength: {path.strength_score:.2f}")
            
            # Inverse resource allocation
            allocations = self.allocate_resources_inverse()
            print("\nInverse Resource Allocation (scrutiny to weaker paths):")
            for i, alloc in allocations.items():
                status = "🔍 HIGH SCRUTINY" if alloc > 0.75 else "OK"
                print(f"  Path {i}: {alloc:.2f} {status}")
            
            print(f"\nConvergence Check: {converged} | Conclusion: {conclusion} | Confidence: {confidence:.2f}")
            
            if converged and confidence > 0.75:
                print("\n✅ CONVERGENCE ACHIEVED - Reasoning is robust\n")
                break
            elif iteration == self.max_iterations - 1:
                print("\n⚠️  Weak convergence or divergence - reasoning needs review\n")
        
        return {
            "problem": self.problem,
            "paths": [
                {
                    "name": p.name,
                    "conclusion": p.conclusion,
                    "strength": p.strength_score,
                    "iterations": p.iterations
                }
                for p in self.paths
            ],
            "converged": converged,
            "final_conclusion": conclusion,
            "confidence": confidence,
            "recommendation": "ACCEPT" if confidence > 0.75 else "REVIEW"
        }


def main():
    """Demonstrate LLM Reasoning Robustness"""
    
    # Example 1: A complex reasoning problem
    problem1 = """
    A company needs to reduce costs by 20%. They have three options:
    A) Cut staff by 15%
    B) Automate processes (high upfront cost, saves in long term)
    C) Negotiate vendor discounts
    
    Which option should they choose?
    """
    
    investigator = LLMReasoningRobustness(problem1)
    result = investigator.run_investigation()
    
    print("\n" + "="*70)
    print("FINAL RESULT")
    print("="*70)
    print(json.dumps(result, indent=2))
    print(f"\nRecommendation: {result['recommendation']}")
    print(f"Act on this conclusion? {'YES - reasoning is robust' if result['confidence'] > 0.75 else 'NO - needs further review'}")


if __name__ == "__main__":
    main()

File: cpp/examples/llm_reasoning_robustness.cpp
#include "../src/amoeba_solver.h"
#include <iostream>
#include <vector>
#include <string>

// ============================================================================
// LLM REASONING ROBUSTNESS: Multi-Path Convergence Analysis
//
// Uses the Amoeba Algorithm to verify LLM reasoning quality:
// - Generate multiple independent reasoning paths for the same problem
// - Evaluate each path's internal consistency and strength
// - Use INVERSE resource allocation (give more scrutiny to weaker paths)
// - Check for convergence across paths → high confidence in reasoning
//
// This addresses: "How do I know my LLM's answer is actually robust?"
// Answer: When multiple independent reasoning approaches converge,
//         that's a strong signal of robustness.
// ============================================================================

Problem setup_llm_reasoning_robustness() {
    Problem problem;
    problem.title = "LLM Reasoning Robustness Analysis";
    problem.description = "Verify LLM answer quality through multi-path reasoning";
    problem.goal = "Achieve convergence across multiple independent reasoning paths";

    // ===== OPTIONS (Reasoning conclusions) =====
    // In this example, we're testing whether an LLM's conclusion "C" is robust
    problem.options = {
        Option{"conclusion_a", "Conclusion A",
               "Derived from step-by-step reasoning",
               "Direct logical progression from premises", 35},

        Option{"conclusion_b", "Conclusion B",
               "Derived from devil's advocate approach",
               "Refutation of opposing views", 40},

        Option{"conclusion_c", "Conclusion C",
               "Derived from constraint-based analysis",
               "Solution satisfying all constraints", 50},

        Option{"conclusion_d", "Conclusion D",
               "Derived from evidence-first analysis",
               "Supported by primary evidence", 45}
    };

    // ===== DATA POINTS (Reasoning quality indicators) =====

    problem.all_data.push_back(DataPoint{
        "path_step_by_step",
        "Step-by-step reasoning leads to C",
        "Logical progression through clear stages converges on C",
        {"conclusion_c"},
        {},
        5
    });

    problem.all_data.push_back(DataPoint{
        "path_devils_advocate",
        "Devil's advocate refutation supports C",
        "Opposing view fails under scrutiny, C remains standing",
        {"conclusion_c"},
        {},
        5
    });

    problem.all_data.push_back(DataPoint{
        "path_constraint_based",
        "Constraint satisfaction points to C",
        "All constraints analyzed; only C satisfies all requirements",
        {"conclusion_c"},
        {},
        5
    });

    problem.all_data.push_back(DataPoint{
        "path_evidence_first",
        "Evidence-first analysis supports C",
        "Primary evidence and data signals point to C",
        {"conclusion_c"},
        {"conclusion_a"},
        4
    });

    problem.all_data.push_back(DataPoint{
        "consistency_measure",
        "All paths show internal consistency",
        "No logical contradictions detected within reasoning chains",
        {"conclusion_c", "conclusion_b"},
        {},
        4
    });

    problem.all_data.push_back(DataPoint{
        "convergence_signal",
        "Multiple independent paths converge",
        "When weak paths are strengthened through iteration, all point to C",
        {"conclusion_c"},
        {"conclusion_a", "conclusion_b"},
        5
    });

    problem.all_data.push_back(DataPoint{
        "strength_reinforcement",
        "Weaker paths strengthen when scrutinized",
        "Inverse resource allocation forces weak reasoning to improve or fail",
        {"conclusion_c"},
        {},
        4
    });

    problem.all_data.push_back(DataPoint{
        "reasoning_robustness",
        "Answer robust under multi-path analysis",
        "Convergence across independent reasoning approaches signals robustness",
        {"conclusion_c"},
        {},
        5
    });

    return problem;
}

// ============================================================================
// MAIN: LLM Reasoning Robustness Verification
// ============================================================================

int main() {
    try {
        std::cout << "\n╔═══════════════════════════════════════════════════════════════════╗\n";
        std::cout << "║                                                                   ║\n";
        std::cout << "║      LLM REASONING ROBUSTNESS: Multi-Path Convergence Analysis    ║\n";
        std::cout << "║                    Using the Amoeba Algorithm                     ║\n";
        std::cout << "║                                                                   ║\n";
        std::cout << "╚═══════════════════════════════════════════════════════════════════╝\n";

        std::cout << "\n📊 SCENARIO:\n";
        std::cout << "An LLM has proposed an answer to a complex reasoning question.\n";
        std::cout << "We verify robustness by testing it through multiple independent\n";
        std::cout << "reasoning paths:\n\n";
        std::cout << "  1. Step-by-step logical progression\n";
        std::cout << "  2. Devil's advocate (refute the opposite)\n";
        std::cout << "  3. Constraint-based analysis\n";
        std::cout << "  4. Evidence-first evaluation\n\n";
        std::cout << "If all paths CONVERGE on the same answer → HIGH CONFIDENCE\n";
        std::cout << "We use INVERSE resource allocation to scrutinize weaker paths.\n\n";

        Problem problem = setup_llm_reasoning_robustness();
        AmoebaInvestigation solver(problem);
        solver.solve();

        std::cout << "\n📋 INTERPRETATION:\n";
        std::cout << "If the algorithm converges on a single conclusion, that reasoning\n";
        std::cout << "is robust. If paths diverge, the LLM's answer needs human review.\n\n";

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }

    return 0;
}
